src/pybind/mgr/rbd_support/module.py

```diff
@@ -483,13 +499,687 @@ def get_perf_counters(self, pool_spec, sort_by):
+                break
+
+        self.sequence += 1
+        task = Task(self.sequence, task_id, action, message, refs)
+
+        # add the task to the rbd_task omap
+        task_json = task.to_json()
+        omap_keys = (task.sequence_key, )
+        omap_vals = (str.encode(task_json), )
+        self.log.info("adding task: {} {}".format(omap_keys[0], omap_vals[0]))
+
+        with rados.WriteOpCtx() as write_op:
+            ioctx.set_omap(write_op, omap_keys, omap_vals)
+            ioctx.operate_write_op(write_op, RBD_TASK_OID)
+        self.append_task(task)
+
+        self.condition.notify()
+        return task_json
+
+    def remove_task(self, ioctx, task, remove_in_memory=True):
+        self.log.info("remove_task: task={}".format(str(task)))
+        omap_keys = (task.sequence_key, )
+        try:
+            with rados.WriteOpCtx() as write_op:
+                ioctx.remove_omap_keys(write_op, omap_keys)
+                ioctx.operate_write_op(write_op, RBD_TASK_OID)
+        except rados.ObjectNotFound:
+            pass
+
+        if remove_in_memory:
+            try:
+                del self.tasks_by_id[task.task_id]
+                del self.tasks_by_sequence[task.sequence]
+            except KeyError:
+                pass
+
+    def execute_task(self, sequence):
+        task = self.tasks_by_sequence[sequence]
+        self.log.info("execute_task: task={}".format(str(task)))
+
+        try:
+            with self.open_ioctx((task.refs[TASK_REF_POOL_NAME],
+                                  task.refs[TASK_REF_POOL_NAMESPACE])) as ioctx:
+                execute_fn = {TASK_ACTION_FLATTEN: self.execute_flatten,
+                              TASK_ACTION_REMOVE: self.execute_remove,
+                              TASK_ACTION_TRASH_REMOVE: self.execute_trash_remove,
+                              TASK_ACTION_MIGRATION_EXECUTE: self.execute_migration_execute,
+                              TASK_ACTION_MIGRATION_COMMIT: self.execute_migration_commit,
+                              TASK_ACTION_MIGRATION_ABORT: self.execute_migration_abort
+                              }.get(task.action)
+                if not execute_fn:
+                    self.log.error("Invalid task action: {}".format(tasks.action))
+                else:
+                    task.in_progress = True
+                    self.in_progress_task = task
+                    self.update_progress(task, 0)
+
+                    self.lock.release()
+                    try:
+                        execute_fn(ioctx, task)
+                    finally:
+                        self.lock.acquire()
+
+                    task.in_progress = False
+                    self.in_progress_task = None
+
+                    self.complete_progress(task)
+                    self.remove_task(ioctx, task)
+
+        except (rados.Error, rbd.Error) as e:
+            self.log.error("execute_task: {}".format(e))
+            self.update_progress(task, 0)
```

Comment:
I am still in process of reading it thus I probably just miss this but how does the user may see if the task completed successfully or with an error?

Reply:
I agree that this is a definite issue right now. The "ceph progress" command shows all running and completed RBD background tasks, but the current progress API doesn't allow you to update the message. Perhaps I'll tweak that API to support that so that I can record failures for completed commands.

Reply:
@dillaman this issue also popped up when I implemented the integration with ceph-dashboard. 
If you're going to make changes to this part of the progress module, here are the dashboard requirements:
* indication of success/failure (mandatory)
* failure description (in case of failure, optional)

I also plan to do another change to the progress module. Currently the event name is just a description, which is not easily parsable. In the dashboard, our async tasks are always identified by a name of the form `<component_name>/<action>`, e.g., `rbd/create`, `rbd/remove`, etc... I would like to add this kind of "convention" to the progress module so that I can parse the event type, and then I can also process the `event.refs` data structure (since it depends on the event type). Do you have any objections to this?

Reply:
@rjfd Couldn't you either use the `task_id` or the `refs` field in the `progress` module to stash whatever data you need to link the two?

Reply:
@rjfd FYI -- it looks like this PR breaks the new dashboard progress service since it runs `dict(ev['refs'])` but I am only storing `["rbd_support"]` in the `refs`. The orchestrator also only stores `["orchestrator"]` in the `refs`

Reply:
> @rjfd Couldn't you either use the `task_id` or the `refs` field in the `progress` module to stash whatever data you need to link the two?

`task_id` (or `Event.id` inside the progress module) is supposed to be a unique identifier of the task/event, hence is usually initialized with a UUID. What I need is a task `type`, and `refs` depends on the task type, i.e., I can only look inside `refs` and understand what's in there if I know the type of the task.

> it looks like this PR breaks the new dashboard progress service since it runs dict(ev['refs']) but I am only storing ["rbd_support"] in the refs

Could you instead pass more information in there, like we do in the dashboard (https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/controllers/rbd.py#L255). In dashboard for an `rbd/delete` operation we always pass a dict with `{'pool_name': <pool_name>, 'image_name': <image_name>}`. This way we can precisely pinpoint which RBD image is being deleted and show some feedback to the user through the UI.
For instance, the final outcome that I imagine is a user to start an RBD delete image task in the CLI and see in the dashboard UI the image row of the RBD table to become gray and saying that the image is being deleted.

Reply:
BTW, in the dashboard `refs` is called `metadata`.

Reply:
@rjfd It's not just this PR, it's the orchestrator module, the OSD PG recovery progress, etc.

Reply:
I know :-) It just happened that you were also touching this part.
Maybe I could open a PR with my proposal and we can all discuss it there?

In the meantime, to unblock the dashboard breakage you add this patch:
`'metadata': dict(ev['refs']),` -> `'metadata': dict(ev['refs']) if ev['refs'] and isinstance(ev['refs'][0], tuple) else {},`

Reply:
@rjfd Sounds good, thx.

Reply:
@rjfd I appended a couple commits to add `progress` module support for failed events and I now pass a `dict` to the `ev.refs` field.